src/control/family.lean

```diff
@@ -0,0 +1,563 @@
+/-
+Copyright (c) 2019 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Simon Hudon
+-/
+import category_theory.category
+import category_theory.types
+import category_theory.pi.basic
+import category_theory.limits.pi
+import category_theory.limits.shapes.binary_products
+import category_theory.limits.shapes.types
+import logic.relation
+
+/-!
+Indexed type families and their categorical structure.
+
+Features:
+
+`fam I`      : family of types and its instance of `category`
+
+Also, support functions for operating with n-tuples of types, such as:
+
+`append1 α β`    : append type J-indexed `β` family to I-indexed `α` family to obtain a `(I⊕J)`-indexed family
+`drop α`         : drop the right component of a `(I⊕J)`-indexed family
+`last α`         : take the right component of a `(I⊕J)`-indexed family
+`append_fun f g` : appends two families of functions `f` and `g`
+`drop_fun f`     : drops the right function family from a `(I⊕J)`-indexed family
+`last_fun f`     : returns the right function family of a `(I⊕J)`-indexed family
+
+Since e.g. `append1 α.drop α.last` is propositionally equal to `α` but not definitionally equal
+to it, we need support functions and lemmas to mediate between constructions.
+-/
+
+universes u v w
+
+open category_theory
+
+/-- Type family indexed by `I`; we call `fam (I ⊕ J)` the product
+of `fam I` and `fam J` (despite being formulated using a sum type) -/
+@[derive inhabited] def fam (I : Type u) := I → Type u
+
+instance {I} : has_one (fam I) :=
+⟨ λ _, punit ⟩
+
+namespace fam
+
+variables {I J : Type u}
+
+/-- retain the left side of a product family -/
+def drop : fam (I ⊕ J) → fam I :=
+λ x i, x (sum.inl i)
+
+/-- retain the right side of a product family -/
+def last : fam (I ⊕ J) → fam J :=
+λ x i, x (sum.inr i)
+
+/-- combine two families with different indices their product -/
+def append1 (f : fam I) (g : fam J) : fam (I ⊕ J)
```

Comment:
Why is this called `append1`?

Reply:
It's a remnant of the type vector constructions. I can rename it

Reply:
I could rename it to `append` but maybe it would be clearer to draw the attention to the similarity to product categories. What do you think @semorrison?

Reply:
The names `drop`, `last` and `append1` don't really make sense in this formulation. How about `left`, `right`, `pair`?